src/_pytest/cacheprovider.py

```diff
@@ -53,7 +54,7 @@
 
 
 @final
-@attr.s
+@attr.s(init=False)
```

Comment:
can you do this with
```
_ispytest: bool = attr.ib(converter=check_ispytest, repr=False)
```

Reply:
This will actually keep a `_ispytest` attribute but I only want an `__init__` argument.

Reply:
I really don't like having to define `__init__` methods in `@attr.s` classes:

perhaps a decorator can work here?

```python
def for_pytest_only(cls: T) -> T:
    original = cls.__init__

    @functools.wraps(original):
    def new_init(self, *args, **kwargs, *_is_pytest=False) -> None:
        if not _is_pytest:
            warnings.warn(...)
        original(*args, **kwargs)

    cls.__init__ = new_init
    return cls

T_Ctor = TypeVar["T_Ctor", bound=Callable[..., T])

def construct_from_pytest(cls: T_Ctor) -> T_Ctor:
    return functools.partial(cls, _is_pytest=True)
```


```python
@for_pytest_only
@attr.s
class Foo:
    ham = attr.s()


_Foo = construct_from_pytest(Foo)
```

Reply:
It should be possible to use a `_is_pytest = attr.ib()` to skip assigning an attribute:

> To be clear: the example above is totally possible with attrs, it's just a bit more clunky.

https://github.com/python-attrs/attrs/issues/342

Reply:
This seems complicated to me. The `__init__` adds a bit of repetition but it is pretty straightforward and easy to understand.

BTW, my own *real* preference is to entirely avoid attrs/dataclasses in any public API, because then it essentially becomes a part of the contract. See also in another language: https://jakewharton.com/public-api-challenges-in-kotlin/.

Reply:
> The `__init__` adds a bit of repetition

avoiding this repetition is the main purpose of `attr.s`, so I'd really rather avoid using `init=False` here

Reply:
is also reduces the very heavy repetition of ` *, _ispytest: bool = False`

Reply:
@graingert The decorator seems too heavy to me, so I prefer not to have it. And init-only arguments are not supported by attrs from what I understand. So I think `init=False` is the way to go for now. So I'm marking this as resolved. But please let me know if you strongly disagree.